Modules/POSIX/Sources/POSIX/Regex.swift

```diff
@@ -0,0 +1,319 @@
+    let description: String
+
+    static func error(from result: Int32, preg: inout regex_t) -> RegexError {
+        var buffer = [Int8](repeating: 0, count: Int(BUFSIZ))
+        regerror(result, &preg, &buffer, buffer.count)
+        let description = String(cString: buffer)
+        return RegexError(description: description)
+    }
+}
+
+
+public final class Regex {
+
+    public struct Options : OptionSet {
+        public let rawValue: Int32
+
+        public init(rawValue: Int32) {
+            self.rawValue = rawValue
+        }
+
+        public static let basic            = Options(rawValue: 0)
+        public static let extended         = Options(rawValue: 1)
+        public static let caseInsensitive  = Options(rawValue: 2)
+        public static let newLineSensitive = Options(rawValue: 4)
+        public static let resultOnly       = Options(rawValue: 8)
+    }
+
+    public struct MatchOptions : OptionSet {
+        public let rawValue: Int32
+
+        public init(rawValue: Int32) {
+            self.rawValue = rawValue
+        }
+
+        public static let firstCharacterNotAtBeginningOfLine = MatchOptions(rawValue: REG_NOTBOL)
+        public static let lastCharacterNotAtEndOfLine        = MatchOptions(rawValue: REG_NOTEOL)
+    }
+
+
+    var preg = regex_t()
+
+    /// Constructs a Regex using the given string pattern and `RegexOptions`.
+    ///
+    /// - parameter pattern: string holding the Regex pattern.
+    /// - parameter options: Options such as `Regex.Options.basic` (BRE) or `Regex.Options.extended` (ERE).
+    ///
+    /// - throws: `RegexError` if the given pattern is not a valid Regex.
+    public init(pattern: String, options: Options = .extended) throws {
+        let result = regcomp(&preg, pattern, options.rawValue)
+
+        guard result == 0 else {
+            throw RegexError.error(from: result, preg: &preg)
+        }
+    }
+
+    deinit {
+        regfree(&preg)
+    }
+
+    /// In the context of a String UTF8View, returns the UTF8View.Index
+    /// of a given character (designated by `regoff`)
+    /// 
+    /// Helper that simplifies the code written at the call site.
+    ///
+    /// - parameter offset:   Offset to a char (returned by native `regexec()`).
+    /// - parameter utf8view: UTF8View of the string.
+    ///
+    /// - returns: UTF8View.Index of the character designated by `regoff`.
+    func index(from offset: regoff_t, in utf8view: String.UTF8View) -> String.UTF8View.Index {
+        return utf8view.index(utf8view.startIndex, offsetBy: Int(offset))
+    }
+
+    /// Check if a given string matches the regex.
+    ///
+    /// - parameter string:  String to validate against the regex.
+    /// - parameter options: Matching options like `firstCharacterNotAtBeginningOfLine` (`REG_NOTBOL` in POSIX parlance).
+    ///
+    /// - returns: `true` if the string matches the regex, else `false`.
+    public func matches(_ string: String, options: MatchOptions = []) -> Bool {
```

Comment:
Just had an idea. What do you think about taking the `public` methods out of the `Regex` type and move them to a `String` extension receiving the `Regex` type as a parameter? This way the call site would become like this:

```swift
let regex = try Regex("hello")
"hello".matches(regex) // true
```
```swift
let regex = try Regex("(hello) (world)")
"hello world".groupsMatching(regex) // ["hello", "world"]
```
```swift
let regex = try Regex("world")
"hello world".replace(regex, with: "zewo") // "hello zewo"
```